AdvLoggerPkg/Library/AdvancedLoggerLib/Dxe/GoogleTest/AdvancedLoggerDxeLibGoogleTest.cpp

```diff
@@ -0,0 +1,297 @@
+/** @file AdvancedLoggerDxeLibGoogleTest.cpp
+
+    This file contains the unit tests for the Advanced Logger DXE Library.
+
+    Copyright (c) Microsoft Corporation.
+    SPDX-License-Identifier: BSD-2-Clause-Patent
+**/
+
+#include <Library/GoogleTestLib.h>
+#include <GoogleTest/Library/MockUefiBootServicesTableLib.h>
+#include <GoogleTest/Protocol/MockAdvancedLogger.h>
+// #include <GoogleTest/Protocol/MockDebugPort.h> // Waiting on mu_basecre DebugPortProtocol mock
+
+extern "C" {
+  #include <Uefi.h>
+  #include <Library/BaseLib.h>
+  #include <Library/DebugLib.h>
+  #include <PiDxe.h>
+  #include <AdvancedLoggerInternal.h>
+  #include <Protocol/DebugPort.h> // Waiting on mu_basecre DebugPortProtocol mock
+  #include "../../AdvancedLoggerCommon.h"
+
+  extern ADVANCED_LOGGER_PROTOCOL  *mLoggerProtocol;
+  extern BOOLEAN                   mInitialized;
+}
+
+using namespace testing;
+
+/**
+  Test class for AdvancedLoggerWrite Function
+**/
+class AdvancedLoggerWriteTest : public Test {
+protected:
+  MockUefiBootServicesTableLib gBSMock;
+  MockAdvancedLogger AdvLoggerProtocolMock;
+  UINTN DebugLevel;
+  CHAR8 *Buffer;
+  UINTN NumberOfBytes;
+  EFI_GUID gMockAdvLoggerProtocolGuid =
+  { 0x434f695c, 0xef26, 0x4a12, { 0x9e, 0xba, 0xdd, 0xef, 0x00, 0x97, 0x49, 0x7c }
+  };
+
+  virtual void

```

Comment:
For my learning purpose, what does `virtual` do here?

Reply:
`virtual` expects the function to be overridden by derived classes, so this keyword was mistakenly copied from the Test class definition. I fixed it so it uses "override" instead.